### DISCUSSION TITLE
How does Lerna detect packages?

### DISCUSSION THREAD

**Discussion by christianmalek** [lerna/lerna#3685] [Q&A] [answered]
Hi,

I'm using Lerna for the first time because it's used by some example packages. The folder hierarchy looks like the following:

```
foo
foo/lerna.json
foo/package.json
foo/package1 (with package.json)
foo/package2 (with package.json)
foo/package3 (with package.json)
```

I'm also needing a server for *foo* and wanted to just include another Lerna repository into the foo project like this:

```
foo/server/lerna.json
foo/server/package.json
foo/server/package1 (with package.json)
foo/server/package2 (with package.json)
foo/server/package3 (with package.json)
```

If I now start a console and my working directory is `foo/` all workspaces defined in `foo/package.json` are getting considered and the scripts like `yarn` work fine.

But if I change to `foo/server` and run scripts like `yarn` Lerna doesn't detect the packages defined in `foo/server/package.json`.

If I instead move the `foo/server` into another folder outside of `foo` it works like a charme.

My question is: Doesn't Lerna consider the working directory I'm in as the root-directory and tries to find a "parent" package?

**Comment by fahslaj** [comment]
For new projects, Lerna uses the workspaces property in the package.json in the root directory (the one with lerna.json in it) to search for packages. For details on this property, see the [npm documentation](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#workspaces) or the [Yarn documentation](https://classic.yarnpkg.com/lang/en/docs/workspaces/). You can tell if Lerna is configured to use this by looking for `"useWorkspaces": true` `lerna.json`.

If you are using pnpm, you might have set npmClient to pnpm in lerna.json. In this case, Lerna will use the packages property in pnpm-workspace.yaml to search for packages. For details on this property, see the [pnpm documentation](https://pnpm.io/workspaces).

If you are using an older version of Lerna or have explicitly opted out of using workspaces ("useWorkspaces" is missing or set to false in `lerna.json`), then Lerna will use the packages property in lerna.json to search for packages.

**Comment by fahslaj** [comment, marked as answer]
You can definitely add server projects to the same lerna repo as your client ones. One example structure would be like this:


```
foo/client/package-1 (with package.json)
foo/client/package-2 (with package.json)
foo/client/package-3 (with package.json)

foo/server/package-4 (with package.json)
foo/server/package-5 (with package.json)
foo/server/package-6 (with package.json)

foo/package.json (this is the root package.json with the "workspaces" property)
foo/lerna.json
```

Your package.json's workspaces property would be set to `["client/*", "server/*"]` and the "useWorkspaces" property in lerna.json would be set to true.
